flocker/ca/_ca.py

```diff
@@ -0,0 +1,225 @@
+            return self.keypair.dump() == other.keypair.dump()
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def selfSignedCACertificate(self, dn, request, serial, expiry, digest):
+        """
+        Sign a CertificateRequest with extensions for use as a CA certificate.
+
+        This code based on ``twisted.internet.ssl.KeyPair.signRequestObject``
+
+        :param DistinguishedName dn: The ``DistinguishedName`` for the
+            certificate.
+
+        :param CertificateRequest request: The signing request object.
+
+        :param int serial: The certificate serial number.
+
+        :param int expiry: Number of seconds from now until this certificate
+            should expire.
+
+        :param str digest: The digest algorithm to use.
+        """
+        req = request.original
+        cert = crypto.X509()
+        dn._copyInto(cert.get_issuer())
+        cert.set_subject(req.get_subject())
+        cert.set_pubkey(req.get_pubkey())
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(expiry)
+        cert.set_serial_number(serial)
+        cert.add_extensions([
+            crypto.X509Extension("basicConstraints", True,
+                                 "CA:TRUE, pathlen:0"),
+            crypto.X509Extension("keyUsage", True,
+                                 "keyCertSign, cRLSign"),
+            crypto.X509Extension("subjectKeyIdentifier", False, "hash",
+                                 subject=cert),
+        ])
+        cert.add_extensions([
+            crypto.X509Extension(
+                "authorityKeyIdentifier", False,
+                "keyid:always", issuer=cert
+            )
+        ])
+        cert.sign(self.keypair.original, digest)
+        return Certificate(cert)
+
+
+def generate_keypair():
+    """
+    Create a new key pair.
+    """
+    return FlockerKeyPair(
+        keypair=KeyPair.generate(crypto.TYPE_RSA, size=4096)
+    )
+
+
+class CertificateAuthority(PRecord):
+    """
+    A certificate authority whose configuration is stored in a specified
+    directory.
+
+    :ivar FilePath path: A ``FilePath`` representing the absolute path of
+        a directory containing the CRT and KEY files.
+    :ivar Certificate certificate: A signed certificate, populated only by
+        loading from path.
+    :ivar FlockerKeyPair keypair: A private/public keypair, populated only by
+        loading from path.
+    """
+    certificate = field(mandatory=True, initial=None)
+    keypair = field(mandatory=True, initial=None)
+    path = field(mandatory=True)
+
+    @classmethod
+    def from_path(cls, path):
+        """
+        :param FilePath path: Directory where private key and certificate are
+            stored.
+        """
+        if not path.isdir():
+            raise PathError(
+                b"Path {path} is not a directory.".format(path=path.path)
+            )
+
+        certPath = path.child("cluster.crt")
```

Comment:
If these are the fixed certificate file names then make them module variables so that they can be shared in the code below.